pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -159,23 +182,77 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error {
 	logger := logging.FromContext(ctx)
 
-	// fetch the equivelant task for this task Run
-	taskName := tr.Spec.TaskRef.Name
-	task, err := c.taskLister.Tasks(tr.Namespace).Get(taskName)
-	if err != nil {
-		logger.Errorf("Failed to reconcile TaskRun: %q failed to Get Task: %q", tr.Name, taskName)
-		return err
+	// TODO(aaron-prindle) verify that build does not have to exist to create a tracker
+	buildRef := corev1.ObjectReference{
+		APIVersion: "build.knative.dev/v1alpha1",
+		Kind:       "Build",
+		Namespace:  tr.Namespace,
+		Name:       tr.Name,
 	}
-	// fetch the build that should exist for the current task run
-	buildSpec := task.Spec.BuildSpec
-	if len(buildSpec.Steps) == 0 && buildSpec.Template.Name == "" {
-		logger.Errorf("Failed to reconcile TaskRun: %q failed to Get BuildSpec: %q", tr.Name, taskName)
+	if err := c.tracker.Track(buildRef, tr); err != nil {
+		logger.Errorf("build %s for taskrun %s: %v", buildRef, tr.Name, err)
 		return err
 	}
-	// TODO use build tracker to fetch build
 
-	// check status of build and update status of taskRun
+	// build creation and status update adapted from prow:
+	// https://github.com/kubernetes/test-infra/blob/8faa040989cf8552683dcdd4400505ca62a3c408/prow/cmd/build/controller.go#L207
+	var wantBuild bool
+
+	// TODO(aaron-prindle) understand how DeletionTimestamp gets set in the first place or if it does...
+	if tr.DeletionTimestamp == nil {
+		wantBuild = true
+	}
+
+	var haveBuild bool
 
+	// we are naming the Build the same as the TaskRun for 1:1 mapping and retrieval
+	b, err := c.getBuild(tr.Namespace, tr.Name)
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return fmt.Errorf("retrieving build %s for taskRun %s: %v", tr.Name, tr.Name, err)
+	case b.DeletionTimestamp == nil:
+		haveBuild = true
+	}
+
+	// Should we create or delete this build?
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			logger.Infof("deleted %s", tr.Name)
+			return nil
+		}
+		switch ownerref := metav1.GetControllerOf(b); {
+		case ownerref == nil, ownerref.APIVersion != groupVersionKind.GroupVersion().String(), ownerref.Kind != groupVersionKind.Kind:
+			logger.Infof("build not controlled by task %s", tr.Name)
+			return nil
+		}
+		logger.Infof("delete builds %s", tr.Name)
+		if err = c.deleteBuild(tr.Namespace, tr.Name); err != nil {
+			return fmt.Errorf("delete build: %v", err)
+		}
+		return nil
+	case tr.Status.Conditions[0].Status != corev1.ConditionUnknown:
```

Comment:
This will panic if the array is empty which what happened when I tried to run it

Reply:
fixed now, thanks!